Insert note dialog: match notes by parent item

abaevbog opened the pull request:
- allow to filter notes by their parent item
- display notes in list mode nested under their parent item
- in library mode, add parent title to the item card

Fixes #2812

<img width="800" height="498" alt="Screenshot 2025-09-19 at 12 23 13 PM" src="https://github.com/user-attachments/assets/e40773d5-0128-45d3-bf6d-b57468d79b90" />
<img width="802" height="504" alt="Screenshot 2025-09-19 at 12 23 32 PM" src="https://github.com/user-attachments/assets/42a0d9ac-2967-4284-ad05-a23b9eed78b6" />

yexingsha commented:
Yes, this looks like a good start!

As for displaying note in item trees, I'm thinking of something similar to annotations: title in regular font size, followed by content in smaller font size, spanning across all columns. This will apply to all item trees everywhere. The only other relevant note fields I can think of are Date Added and Date Modified, but they are probably less useful than simply displaying more note content.

adomasven commented:
Date Added and Modified are actually probably important enough to keep, especially for standalone notes where that's their only other identifier besides note content.

yexingsha commented:
A read only note editor side pane sounds good to me! I don't think it would be too much, especially since we are using the side pane for annotations anyway.

yexingsha commented:
I think we can slightly increase the width of the note preview area, but otherwise this looks good!

abaevbog commented:
Sure, no reason not to make it a bit wider. I bumped the width to 250px:

<img width="888" height="499" alt="Screenshot 2025-09-26 at 12 10 33 PM" src="https://github.com/user-attachments/assets/b535ebd7-11c5-46c8-8c4f-99f4798570be" />

abaevbog committed:
citation dlg: insert note refactoring

- allow one to insert multiple notes at a time
- in list mode, match notes by their parent item and
render them as children of collapsible regular items
- in library mode, add parent title to the item card
similar to annotation and add note preview sidebar that
displays the content of the note when it is selected

- refactor list mode to use virtualized-table for rendering.
Now that we need to render parent-child row relations,
virtualized-table offers more flexibility, and we can no longer
have the same workflow for suggested items in library mode as for list mode.

abaevbog commented:
@yexingsha, I rebased this PR that we worked on a while back, now that "Add annotations" is merged. It also has an additional tweak: you can now select multiple notes as bubble before inserting them all together. Bubbles can be reordered, same as with citations. Do you mind having another look to make sure it looks right?

(In this PR I also refactored list mode to use virtualized-table instead of sharing DOM nodes generation with library mode. Now that we want collapsible item containers with child nodes, that sharing no longer really works. Also virtualized-table will be necessary anyways for potential future list mode of "Add annotation" dialogs , e.g. for mobile, so I thought might as well do it now. All of that is to say that the way list mode works is completely different here but I tried to preserve the look).

yexingsha commented:
The bubble note input looks pretty good to me! Just a few side issues:

- In the note dropdown popover, there needs to be a 4px gap between the item name and the note content, and a 16px gap between the note content and the buttons.
- I don't remember if/when we decided to not show anything in the list mode when there is no selected and open items; currently the missing result section looks pretty weird. This situation can happen in the citation dialog as well, but it will be quite common in the insert note dialog when none of the selected and open items have notes. I vaguely remember an iteration where we had some placeholder text there for this situation? But it might be better if we show the most recently modified/read notes/items instead. We can even consider adding a "recently modified/read" as a persisting section in the list mode, so that it will always show a number of suggestions when there is no active search.
- There is a resizing issue where if the bubbles occupy one line in library mode, but wrap to two lines when switching to list mode, the height calculation will be off and the bottom area will be cut off:

<img width="868" height="154" alt="Screenshot 2026-04-17 at 16 12 32" src="https://github.com/user-attachments/assets/fc2e7052-a02f-4fe4-9a31-e8346203827f" />

abaevbog committed:
fix bottom area cutoff after list mode resize

If bubbles overflow to the second line

abaevbog committed:
fix spacing and info in details popup

abaevbog committed:
fix test breakage due to lazy row rendering in vt

abaevbog committed:
show message in list mode on no search results

abaevbog commented:
With the added "No results" message in list mode, this is what the dialog looks like now when there's nothing to show. What do you think?

<img width="806" height="143" alt="Screenshot 2026-04-24 at 12 33 44 PM" src="https://github.com/user-attachments/assets/cedac65b-6280-4d4e-b234-b1770fd172a3" />
<img width="808" height="149" alt="Screenshot 2026-04-24 at 12 32 28 PM" src="https://github.com/user-attachments/assets/2ecacafa-b5f1-4228-8a9a-1a6bedb38556" />

yexingsha commented:
I think it works, though the wording should be consistent with library mode ("No selected or open notes" versus "No notes are selected"). And when the dialog reloads results (e.g. when switching focus from a different window back to the dialog), the message should stay displayed, just like in library mode.

abaevbog committed:
fix "no search results" blinking on refocus

abaevbog committed:
tweak messages phrasing

abaevbog committed:
fix open child note not showing in library mode

abaevbog commented:
> ("No selected or open notes" versus "No notes are selected")

Thanks for pointing that out. Fixed, we now have "No selected or open notes" and "No selected or open items" everywhere.

> And when the dialog reloads results (e.g. when switching focus from a different window back to the dialog), the message should stay displayed, just like in library mode.

Oh yeah, that was because on refocus we would rerun search which refreshes items list twice: one after quick refresh of selected/open items, and the second after a potentially longer refresh of library search results. The message generally is rendered after second rerun to not flash the message mid-search. But in the case of refocus, we'll render the message on the first run too, which eliminates the blinking.
